### ISSUE TITLE
(elasticloadbalancingv2): #13332 is a breaking change!

### ISSUE THREAD

**Issue by akefirad** [aws/aws-cdk#13437]
## :question: General Issue

#13332 is a breaking change!

### The Question
I believe #13332 fixing (#13150) is a breaking change since it adds the resource (with certificates) and then remove the old one (with same certificate), which results to removing the existing certificates.

I can't think of anything other than, doing the migration is two deployment or adding the missing certificates manually after the deployment. Both causes some disruption to the service (endpoints are gonna be unavailable for a couple of seconds/minutes).

I'd suggest to update the release note accordingly.

**Comment by akefirad** [comment]
Confirmed. The first certificates in the list will be gone, if one upgrades to version 1.92.0.

**Comment by njlynch** [comment]
Some potential learnings so far:

It appears that as-is, the `AWS::ElasticLoadBalancingV2::ListenerCertificate` resource reacts as expected on creates/deletes, but does ... odd .. things on update.

For the original bugs (#13150 and #13329), the issue was that if the listener was updated from having 2 to 3 certificates (requiring an update of the ListenerCertificate resource), one of the certificates would be dropped. Here, doing an upgrade from v1.91.0 to v1.92.0 means that any users that have successfully deployed with two certs attached to a single `ListenerCertificate` would see an update to that `ListenerCertificate` and the creation of a new `ListenerCertificate`. The creation appears to work, but the update causes one of the certificates to be removed.

-- **Deployed successfully (all at once, no updates) on v1.91.0** --

```
  LBListener49E825B4:
    Type: AWS::ElasticLoadBalancingV2::Listener
    Properties:
      DefaultActions:
        - RedirectConfig:
            Host: example.com
            StatusCode: HTTP_302
          Type: redirect
      LoadBalancerArn:
        Ref: LB8A12904C
      Certificates:
        - CertificateArn: arn:aws:acm:eu-west-1:123456789012:certificate/f53b3399-2092-4fd7-a777-33734a4dad22
      Port: 443
      Protocol: HTTPS
  LBListenerSSLCert177C324A:
    Type: AWS::ElasticLoadBalancingV2::ListenerCertificate
    Properties:
      Certificates:
        - CertificateArn: arn:aws:acm:eu-west-1:123456789012:certificate/b6e4aa65-6d49-4a4e-8321-af2790e4e128
        - CertificateArn: arn:aws:acm:eu-west-1:123456789012:certificate/8f87cdba-816e-4dda-b606-7760e513ca3d
      ListenerArn:
        Ref: LBListener49E825B4
```

-- **v1.92.0** --

```
  LBListener49E825B4:
    Type: AWS::ElasticLoadBalancingV2::Listener
    Properties:
      DefaultActions:
        - RedirectConfig:
            Host: example.com
            StatusCode: HTTP_302
          Type: redirect
      LoadBalancerArn:
        Ref: LB8A12904C
      Certificates:
        - CertificateArn: arn:aws:acm:eu-west-1:123456789012:certificate/f53b3399-2092-4fd7-a777-33734a4dad22
      Port: 443
      Protocol: HTTPS
  LBListenerSSLCert177C324A:
    Type: AWS::ElasticLoadBalancingV2::ListenerCertificate
    Properties:
      Certificates:
        - CertificateArn: arn:aws:acm:eu-west-1:123456789012:certificate/b6e4aa65-6d49-4a4e-8321-af2790e4e128
      ListenerArn:
        Ref: LBListener49E825B4
  LBListenerSSLCert26015FAF3:
    Type: AWS::ElasticLoadBalancingV2::ListenerCertificate
    Properties:
      Certificates:
        - CertificateArn: arn:aws:acm:eu-west-1:123456789012:certificate/8f87cdba-816e-4dda-b606-7760e513ca3d
      ListenerArn:
        Ref: LBListener49E825B4
```

-- **cdk diff** --

```
Stack Issue13437Stack
Resources
[+] AWS::ElasticLoadBalancingV2::ListenerCertificate LB/Listener/SSLCert2 LBListenerSSLCert26015FAF3
[~] AWS::ElasticLoadBalancingV2::ListenerCertificate LB/Listener/SSLCert LBListenerSSLCert177C324A replace
 └─ [~] Certificates (requires replacement)
     └─ @@ -1,8 +1,5 @@
        [ ] [
        [ ]   {
        [ ]     "CertificateArn": "arn:aws:acm:eu-west-1:123456789012:certificate/b6e4aa65-6d49-4a4e-8321-af2790e4e128"
        [-]   },
        [-]   {
        [-]     "CertificateArn": "arn:aws:acm:eu-west-1:123456789012:certificate/8f87cdba-816e-4dda-b606-7760e513ca3d"
        [ ]   }
        [ ] ]
```

In the above, the `b6e4aa65-6d49-4a4e-8321-af2790e4e128` certificate was the one that was dropped/removed entirely by the update.

The code was written to be backwards-compatible, such that if only two certificates were attached to a listener, no logical IDs would change. I suspect that is what needs to change here; if we create a new logical ID for both certificates, both should get "clean" (single-cert) resources, and both should update appropriately.

**Comment by akefirad** [comment]
🤔 I don't think so there's a solution for this. Your suggestion wouldn't solve the issue either (didn't test it through). I suspect it removes both certificates at the last step which is deleting the old resource.

**Comment by njlynch** [comment]
I just emulated the patch on my test stack, and it appears to work:

-- **patch** --

```ts
    const listenerCert1 = listener.node.findChild('SSLCert') as elbv2.ApplicationListenerCertificate;
    (listenerCert1.node.defaultChild as elbv2.CfnListenerCertificate).overrideLogicalId('LBListenerSSLCertOverride1234')
```

-- **cdk diff** --

```
Stack Issue13437Stack
Resources
[-] AWS::ElasticLoadBalancingV2::ListenerCertificate LBListenerSSLCert177C324A destroy
[+] AWS::ElasticLoadBalancingV2::ListenerCertificate LB/Listener/SSLCert LBListenerSSLCertOverride1234
[+] AWS::ElasticLoadBalancingV2::ListenerCertificate LB/Listener/SSLCert2 LBListenerSSLCert26015FAF3
```

By deploying the above, all `ListenerCertificate` operations are Creates/Deletes, which appear to work properly (all 3 certs on my listener were maintained). This is logically equivalent to us changing the logical ID in the code for the first `ListenerCertificate`. I'll spin up a fix.

**Comment by akefirad** [comment]
Interesting! How come `destroy` is not removing the certificates? Effectively it's the same as before; adding new certificates, removing the old resource (which removes the added certificates).

**Comment by akefirad** [comment]
> When the resource is deleted, it is removing the relationship by that ID, rather than by the individual certificate ARNs.
Interestingly I remember the opposite. I forgot the exact details of my experience. But this is what I recall:
1. I had an ALB with two certificates (one main, `foo.com` and one additional, `bar.com`), everything created by CDK.
2. I added an additional certificate `baz.com` to the listener in my CDK code. Deployed the code and one additional certificate (either `bar.com` or `baz.com`, don't remember which one) was gone.
3. I added the missing certificate manually to the listener.
4. Upgraded CDK to v1.92.0 and re-deployed the stack again. The additional certificate was gone (the first one in the list) which was because CFN firsts adds the new resource and then remove the old one, which effectively means the certificate is gone (seems a design flaw to me on CFN side).

Two things that is different with your example/suggestion
1. One certificate was added manually to the listener.
2. The resource is not replaced by removed completely. (I can't see how this might be relevant)

Let me know if I can help with anything.
